ProjectSidewalk/SidewalkWebpage
# 1197 better syncing b/w MTurk and Project Sidewalk

Resolves #1197 
Fixes #1230 
Resolves #1203 

This PR improves how Project Sidewalk keeps track of turkers' work and it adds a timer that is visible to turkers that shows how much longer they are able to audit.

The changes:
* Added a new timer to the navbar that shows how much time the turker has left to audit. It turns red when there are less than 15 minutes remaining.
    ![timer-less-than-15-audit](https://user-images.githubusercontent.com/6518824/49535993-5b85e400-f87a-11e8-94b1-80ead6d6a127.png)
* When the time that a turker is able to audit has run out, the timer says "Expired" and an overlay appears over the audit page preventing them from auditing further. However, they are still able to navigate the rest of the Project Sidewalk website.
    ![timer-expired-audit](https://user-images.githubusercontent.com/6518824/49536155-b7506d00-f87a-11e8-977d-9f3eb8161874.png)
* The timer for a turker now starts when they click the Start HIT button. We used to start the timer when the turker added the assignment to their queue. We accomplish this by sending the number of minutes the turker is allowed to audit as a query parameter in the URL hidden in the Start HIT button.
* The `assignment_end` column in the `amt_assignment` table is now a required column. This column now represents the time when the turker must stop auditing, whereas it used to represent when they stopped auditing (although this was only recorded for the first HIT a turker did with us: #1230)
* The `completed` column of the `amt_assignment` table is now correctly marked as complete for returning turkers after they complete at least one new mission in their new HIT.
* Resuming an assignment is now recorded correctly in the `amt_assignment` table instead of generating a new record.
* The `confirmation_code` column in the `amt_assignment` table is now a required column.

Testing:
1. You can simulate arriving at Project Sidewalk from MTurk by using the URL http://localhost:9000?referrer=mturk&hitId=h1&workerId=worker1&assignmentId=a1&minutes=20 This will create a turker account for you on Project Sidewalk where your turker id (and username) is "worker1" and you will have 20 minutes to work.
1. Whenever I want to test a new account, I just increment all the numbers (so h2, a2, worker2).
1. You should be sent directly to the audit page. There should be a timer in the navbar that is counting down from 20 minutes.
1. Try using the same URL again. You should be logged back in but the counter _will not_ reset. It should continue from where it started the first time.
1. Try the same URL but change the assignment number. Now your timer should reset. This is necessary in case a turker starts a new assignment before their old one has been completed.
1. Wait for the timer to go below 15 minutes. The text should now turn red.
1. Navigate around the rest of the website. The timer should remain in the navbar.
1. Complete the tutorial and your first mission. A confirmation code button should be added on the left side of the interface.
1. Wait for the timer to go to 0 (you might want to just log in as a new turker with a 1 minute long assignment so you don't have to wait). The timer should now say "Expired". If you are on the audit page, a black overlay should show up preventing you from auditing further. You should still be able to navigate to the rest of the website without a problem though. But the black overlay should appear whenever you go back to the audit page.
1. For a worker account where you have already received a confirmation code, create a new assignment (by using the same workerId, but different assignmentId in the URL). You should _not_ have to take the tutorial again. The confirmation code button should no longer be on the interface.
1. Audit a mission in this new assignment. It should add the confirmation code button again, but with a new confirmation code this time.
1. For every assignment where you completed at least one mission, the corresponding row in the `amt_assignment` table should have `completed=TRUE`.
1. Do some more stress testing and see what you can find!
